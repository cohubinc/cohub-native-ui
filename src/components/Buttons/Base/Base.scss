.CohubButton {
  // font-family: Akkurat-Mono;
  font-weight: lighter;
  border: none;
  border-radius: var(--default-border-radius);
  padding: 6px 12px;
  border: none;
  outline: none;
  cursor: pointer;
  color: var(--white);
  transition: all 50ms ease-in-out;
  letter-spacing: 0.05rem;

  .content-container {
    transition: all 50ms ease-in-out;
  }

  &:active {
    box-shadow: none !important;
    filter: brightness(97%);

    .button-text {
      transform: translateY(-1px);
    }
  }

  // &:focus {
  //   filter: brightness(93%);
  // }

  // &:hover {
  //   filter: brightness(97%);
  // }

  &:disabled {
    cursor: default;
    opacity: 0.4;

    &:hover {
      filter: inherit;
    }
  }
}