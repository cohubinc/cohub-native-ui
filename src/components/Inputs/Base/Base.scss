.GenericInput {
  input,
  textarea {
    position: relative;
    padding: 14px 12px;
    width: 100%;
    font-size: 14px;
    display: block;
    border-radius: var(--default-border-radius);
    border: 1px solid var(--divider-grey);

    &:focus ~ .bar:before {
      width: 100%;
    }
  }

  input[type="text"],
  input[type="email"],
  input[type="password"] {
    -webkit-appearance: none;
  }

  input[type="password"] {
    letter-spacing: 0.2rem;
  }

  label {
    color: var(--grey-600);
    font-weight: normal;
    position: absolute;
    pointer-events: none;
    left: 10px;
    top: 12px;
    transition: 100ms ease all;
    border-radius: none;

    &.FloatedLabel {
      top: -10px;
      font-size: 12px;
      color: var(--grey-600);
      padding-left: 4px;
      padding-right: 4px;
    }
  }

  .bar {
    position: relative;
    display: block;
    width: 100%;

    &:before {
      content: "";
      height: 2px;
      width: 0;
      bottom: 0;
      position: absolute;
      background: var(--primary-green);
      transition: 150ms ease all;
    }
  }
}

.ContrastInput {
  input,
  textarea {
    position: relative;
    padding: 14px 12px;
    width: 100%;
    min-height: 40px;
    font-size: 14px;
    display: block;
    border-radius: var(--default-border-radius);
    border: none;

    &:focus ~ .bar:before {
      width: 100%;
    }
  }

  input[type="text"],
  input[type="email"],
  input[type="password"] {
    -webkit-appearance: none;
  }

  input[type="password"] {
    letter-spacing: 0.2rem;
  }

  label {
    color: var(--grey-600);
    font-weight: normal;
    position: absolute;
    pointer-events: none;
    left: 8px;
    top: 12px;
    transition: 100ms ease all;
    border-radius: none;

    &.FloatedLabel {
      top: 1px;
      font-size: 10px;
      color: var(--grey-600);
      // padding-left: 4px;
      // padding-right: 4px;
    }
  }

  .bar {
    position: relative;
    display: block;
    width: 100%;

    &:before {
      content: "";
      height: 2px;
      width: 0;
      bottom: 0;
      position: absolute;
      background: var(--primary-green);
      transition: 150ms ease all;
    }
  }
}
