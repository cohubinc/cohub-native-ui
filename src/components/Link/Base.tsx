import React from "react";
import { Link } from "react-router-dom";
import { TLinkProps } from "./TLinkProps";
import Color from "src/definitions/enums/Color";

export default class Base extends React.PureComponent<TLinkProps> {
  render() {
    const {
      href,
      styled,
      animated,
      className,
      style,
      to,
      onClick,
      ...restProps
    } = this.props;

    const classes =
      styled && animated ? `cohub-link ${className || ""}` : className;

    const linkStyle = {
      color: (styled ? Color.link : "inherit") as any,
      fontSize: "inherit",
      cursor: "pointer",
      ...style
    };

    if (href || onClick || !to) {
      return (
        <a
          className={classes}
          style={linkStyle}
          href={href}
          onClick={onClick}
          {...restProps}
        />
      );
    }

    return (
      <Link
        to={to}
        className={classes}
        style={linkStyle}
        onClick={onClick}
        {...restProps}
      />
    );
  }
}
