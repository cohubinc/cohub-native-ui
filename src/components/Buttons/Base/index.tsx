import React from "react";
import {
  Text,
  Animated,
  LayoutChangeEvent,
  StyleProp,
  ViewStyle,
  TouchableHighlight,
  View,
  ActivityIndicator
} from "react-native";
import IButtonProps from "../../../definitions/interfaces/IButtonProps";
import Colors from "../../../definitions/enums/Color";
import gs from "../../../definitions/constants/GlobalStyles";
import Color from "../../../definitions/enums/Color";
import { Typography } from "src";

export { IButtonProps };

export default abstract class Base<T = {}> extends React.Component<
  IButtonProps<T>
> {
  // Animation to enlarge the container
  _pulseAnimation: Animated.Value;
  _pulseStyle: any;

  // Animation to draw a loader under the text content
  _lineAnimation: Animated.Value;
  _lineStyle: any;
  _lineContainerStyle: StyleProp<ViewStyle>;

  static defaultProps = {
    animated: true,
    color: Colors.primary
  };

  constructor(props: IButtonProps<T>) {
    super(props);

    this._pulseAnimation = new Animated.Value(0);
    this._pulseStyle = props.animated
      ? {
          transform: [
            {
              scale: this._pulseAnimation.interpolate({
                inputRange: [0, 1],
                outputRange: [1, 1.08]
              })
            }
          ]
        }
      : null;

    this._lineAnimation = new Animated.Value(0);
    this._lineStyle = null;
    this._lineContainerStyle = {
      overflow: "hidden",
      width: "100%",
      height: 4,
      bottom: 0,
      position: "absolute"
    };
  }

  componentDidMount() {
    if (this.props.loading) {
      const loop = Animated.loop(
        Animated.timing(this._lineAnimation, {
          toValue: 100,
          duration: 4000,
          useNativeDriver: true
        })
      );

      loop.start();
    }
  }

  componentDidUpdate(prevProps: IButtonProps<T>) {
    const loop = Animated.loop(
      Animated.timing(this._lineAnimation, {
        toValue: 100,
        duration: 4000,
        useNativeDriver: true
      })
    );

    if (!prevProps.loading && this.props.loading) {
      loop.start();
    }

    if (prevProps.loading && !this.props.loading) {
      loop.stop();
    }
  }

  beginPress = () => {
    Animated.timing(this._pulseAnimation, {
      toValue: 1,
      duration: 100,
      useNativeDriver: true
    }).start();
  };

  endPress = () => {
    Animated.timing(this._pulseAnimation, {
      toValue: 0,
      duration: 100,
      useNativeDriver: true
    }).start();
  };

  onLayout = (event: LayoutChangeEvent) => {
    const { animated, color } = this.props;

    const width = event.nativeEvent.layout.width;

    this._lineStyle = animated
      ? {
          left: 0,
          width: "50%",
          height: 4,
          bottom: 0,
          position: "absolute",
          borderRadius: 4,
          backgroundColor: color || Color.primary,
          transform: [
            {
              translateX: this._lineAnimation.interpolate({
                inputRange: [0, 10, 30, 40, 60, 100],
                outputRange: [-0.5, 1.0, 1.5, 1.0, -0.5, -0.5].map(
                  val => val * width * 0.5
                )
              })
            },
            {
              scaleX: this._lineAnimation.interpolate({
                inputRange: [0, 10, 30, 40, 60, 100],
                outputRange: [0, 1.0, 0, 1.0, 0, 0]
              })
            }
          ]
        }
      : null;

    this.forceUpdate();
  };

  render() {
    const {
      labelStyle,
      label,
      color,
      mono,
      bold,
      disabled,
      loading
    } = this.props;

    return (
      <TouchableHighlight
        underlayColor={Color.trueWhite}
        onPressIn={this.beginPress}
        onPressOut={this.endPress}
        onLayout={this.onLayout}
        {...this.props}
      >
        <Animated.View
          style={[
            this.props.style,
            this._pulseStyle,
            { flexDirection: "row", alignItems: "center" }
          ]}
        >
          <Typography
            color={color}
            style={[labelStyle, disabled ? { opacity: 0.3 } : undefined]}
            mono={mono}
            bold={bold}
          >
            {label}
          </Typography>
          {loading && (
            <ActivityIndicator style={{ marginLeft: 8 }} color={color} />
          )}
        </Animated.View>
      </TouchableHighlight>
    );
  }
}
