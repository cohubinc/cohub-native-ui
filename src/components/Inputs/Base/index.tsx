import React from "react";

import HTMLElementProps from "src/definitions/types/HtmlElementProps";
import FloatingLabelWrapper from "../FloatingLabelWrapper";

export interface IProps {
  /**
   * Floating label for the input
   */
  label?: string;

  /**
   * Input is invalid
   */
  error?: boolean;

  appearance?: "contrast" | "inverted";

  /**
   * HTML attribute for debugging the input
   */
  "data-qa"?: string;

  /**
   * HTML attribute for debugging the label
   */
  "data-qa-label"?: string;
}

type TProps = IProps & HTMLElementProps<HTMLInputElement>;

interface IState {
  hasFocus: boolean;
}

export default class Base extends React.PureComponent<TProps, IState> {
  static defaultProps: Partial<TProps> = {
    type: "text",
    autoComplete: "off",
    autoFocus: false,
    "data-qa": "base-input-element",
    "data-qa-label": "base-input-element-label"
  };

  render() {
    const {
      style,
      className,
      appearance,
      label,
      onClick,
      onFocus,
      onBlur,
      onChange,
      value,
      error,
      ...restProps
    } = this.props;

    return (
      <FloatingLabelWrapper
        data-qa-label={this.props["data-qa-label"]}
        {...{
          style,
          className,
          appearance,
          label,
          onClick,
          onFocus,
          onBlur,
          onChange,
          value,
          error
        }}
      >
        {({ componentProps }) => <input {...componentProps} {...restProps} />}
      </FloatingLabelWrapper>
    );
  }
}
