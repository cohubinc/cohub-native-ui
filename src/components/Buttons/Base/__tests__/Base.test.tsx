import React from "react";
import { shallow } from "enzyme";

import Base from "../index";

describe("Base Button", () => {
  it("renders without crashing", () => {
    const wrapper = shallow(<Base>push me</Base>);
    expect(wrapper).toMatchSnapshot();
  });
});
