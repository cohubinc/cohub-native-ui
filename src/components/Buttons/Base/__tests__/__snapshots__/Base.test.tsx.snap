// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Base Button renders without crashing 1`] = `
<button
  className="CohubButton "
  style={
    Object {
      "backgroundColor": "hsl(0, 0%, 77%)",
      "boxShadow": "0 2px 3px hsla(0, 0%, 0%, 0.2), inset 0px 2px 0px hsl(0, 0%, 92%)",
    }
  }
>
  <div
    className="button-text relative"
  >
    <Typography
      color="#EFF7EE"
      secondary={true}
      style={
        Object {
          "opacity": 1,
          "transition": "opacity 150ms ease-in",
        }
      }
      uppercase={true}
    >
      <span>
        push me
      </span>
    </Typography>
  </div>
</button>
`;
